Throw DirectoryNotFoundException on Image.Save with bad directory

Macromullet opened the pull request:
This provides a more meaningful exception when saving an Image to a directory that does not exist. Addresses #31367

This does change behavior slightly per https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/breaking-changes.md#bucket-2-reasonable-grey-area, as previously, the operation would throw a System.Runtime.InteropServices.ExternalException.

Macromullet committed:
Provide better exception when saving to non-existent directory

When saving an image to a non-existent directory, throw a DirectoryNotFoundException
vs a System.Runtime.InteropServices.ExternalException.

Fix #31367

Macromullet committed:
Change exception message to be more intuitive

Use: The directory {0} of the filename {1} does not exist.
instead of: The target directory {0} of the targetPath {1} does not exist.

Fix #31367

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=34998) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=34998) before we can accept your contribution.<br/><br/>:x: Macromullet [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=34998)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=34998) it.</sub>

Macromullet committed:
Fix test on unix by using built path.

We were using a hardcoded path with backslashes. That didn't match the built path for the Assert.

Macromullet committed:
Make Test conditional as it requires GDI+

Macromullet committed:
Put directory check closer to where it's used to ensure that we
capture it from all Save variations.

Macromullet committed:
Remove trailing whitespace

Macromullet commented:
This is working in .net core, but failing net472 tests. Is this because in that mode it's acting as a facade and just redirecting to the .net framework implementation? If so, I don't see a way of making the tests pass across the board, without understanding the compatibility change, as the behavior in .net 472 was to throw the GDI+ exception as well.

danmosemsft commented:
@Macromullet you're exactly right. Some libraries' tests (like this one) test on .NET Framework as well even though we don't update add features to them on .NET Framework any more. This is just to catch inadvertent compat breaks. In this case it's intentional. Please fix it by adding `[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, ".NET Framework throws ExternalException")]` or similar to those test(s).

Macromullet commented:
Thanks @danmosemsft !

Macromullet committed:
Update test to ensure it doesn't run on the .NET Framework

safern commented on src/libraries/System.Drawing.Common/src/System/Drawing/Image.cs:
```diff
@@ -153,6 +153,15 @@ public void Dispose()
         /// </summary>
         public void Save(string filename) => Save(filename, RawFormat);
 
+        private static void CheckDirectoryExists(string filename)
```
Nit: we usually name these methods, `ThrowIfDirectoryDoesntExist` so that when you're reading call sites, you know that method is going to throw if it doesn't exist.

Macromullet replied:
Awesome. Thanks for the feedback. That was my first choice, but other calls were calling CheckStatus(), which actually does a throw, so I tried to align my renaming with that. Do you have any issue with me renaming that method to ThrowIfStatusNotOk(), as well as this one?

safern replied:
Nope, I don't mind at all, it is always good to be consistent.

Macromullet replied:
After checking more closely, there are 590 calls to CheckStatus(), spread amongst 36 files. Given the scope of that change, it probably makes sense to do that in a different issue/PR

safern replied:
Yeah I was going to mention that as well 😄

Macromullet committed:
Change CheckIfDirectoryExists to ThrowIfDirectoryDoesntExist
to match naming conventions for throw-style methods.

Fix #31367

safern commented on src/libraries/System.Drawing.Common/tests/ImageTests.cs:
```diff
@@ -199,5 +199,16 @@ public void GetEncoderParameterList_ReturnsExpected(ImageFormat format, Guid[] e
+
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, ".NET Framework throws ExternalException")]
+        [ConditionalFact(Helpers.IsDrawingSupported)]
+        public void Save_InvalidDirecotry_ThrowsDirectoryNotFoundException()
```
Nit: typo in the name. 
```suggestion
        public void Save_InvalidDirectory_ThrowsDirectoryNotFoundException()
```

Macromullet replied:
Thanks. I should have caught that.

safern commented:
Yeah. I've been seen that across multiple builds, it seems like nuget.org is having some trouble (probably outage) that is hitting us. We can retry the build when the others finish. Anyway you had a typo which will cause you to update the PR and the builds will restart.

Macromullet committed:
Fixes typo in test name

Fix #31367

Macromullet commented:
The pipeline seems to stall out from time to time. Is this normal? Is it a prioritization issue via the number of available agents?

danmosemsft commented:
@Macromullet yes there are some queues that have been a bit low on machines recently.

Macromullet commented:
So the checks failed here:
  Starting:    System.IO.FileSystem.Tests (parallel test collections = on, max threads = 2)
    System.IO.Tests.DirectoryInfo_Create.RootPath_AppContainer [SKIP]
      Condition(s) not met: "IsInAppContainer"
    System.IO.Tests.File_GetSetTimes.TimesIncludeMillisecondPart_HFS [SKIP]
      Condition(s) not met: "isHFS"

That's outside the scope of this change. I'm unsure how to proceed since I don't control that body of code.

safern commented:
That completely seems like an unrelated error. Will retry the build, but will save the links to investigate and open an issue accordingly.

Macromullet commented:
Looks like we are good to go now. Thanks!

safern commented:
Thanks, @Macromullet for the contribution 😄! For the test failure I opened #35117
